server/controllers/user.js

```diff
@@ -55,44 +56,64 @@ export const getUser = async (req, res, next) => {
 }
 
 export const followUser = async (req, res, next) => {
-  if (req.body.userId !== req.params.id) {
-    try {
-      const user = await User.findById(req.params.id)
-      const currentUser = await User.findById(req.body.userId)
-      if (!user.followers.includes(req.body.userId)) {
-        await user.updateOne({ $push: {followers: req.body.userId }})
-        await currentUser.updateOne({ $push: {following: req.params.id }})
-        res.status(200).json("User followed!")
-      } else {
-        return next(createError(403, 'You already follow this user'))
-      }
-    } catch (error) {
-      next(error)
-    }
-  } else {
+  const currentUser = await getCurrentUser(req)
+  const userToFollow = await User.findById(req.params.id);
+
+  if (!userToFollow) {
+    return next(createError(400, 'No user could be found with this ID'))
+  }
+
+  const currentUserId = currentUser._id.toString();
+  const userToFollowId = userToFollow._id.toString();
+
+  if (currentUserId === userToFollowId) {
     return next(createError(403, 'You cannot follow yourself'))
   }
-} 
+
+  if (userToFollow.followers.includes(currentUserId)) {
+    await currentUser.updateOne({ $push: { following: userToFollowId }})
```

Comment:
The idea here is to make sure the `currentUser` is in the correct state. I managed to get myself into a position where `userToFollow.followers` included the id for the `currentUser`, but `currentUser.following` did not include the ID for `userToFollow`. Not sure how, probably while it was only half working. But anyway a bit of rudimentary protection to tidy that up if it happens again!